flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/ContinuousHistogram.scala

```diff
@@ -0,0 +1,337 @@
+        (b - getValue(index)) / (getValue(index + 1) - getValue(index))
+      s = (getCounter(index) + m_b) *
+        (b - getValue(index)) / (2 * (getValue(index + 1) - getValue(index)))
+    }
+    for (i <- 0 to index - 1) {
+      s = s + getCounter(i)
+    }
+    s = s + getCounter(index) / 2
+    s.round.toInt
+  }
+
+  /** Returns the total number of entries in the histogram
+    *
+    * @return Total number of points added in the histogram
+    */
+  override def total: Int = {
+    var result = 0
+    for (i <- 0 to bins - 1) {
+      result = result + getCounter(i)
+    }
+    result
+  }
+
+  /** Bin counter access function
+    *
+    * @param bin bin number to access
+    * @return `m_bin` = counter of bin
+    */
+  private[math] def getCounter(bin: Int): Int = {
+    require(0 <= bin && bin < bins, bin + " not in [0, " + bins + ")")
+    data.apply(bin)._2
+  }
+
+  /** Number of bins currently being used
+    *
+    * @return number of bins
+    */
+  override def bins: Int = {
+    data.size
+  }
+
+  /** Returns the string representation of the histogram.
+    *
+    */
+  override def toString: String = {
+    s"Size:" + bins + " " + data.toString
+  }
+
+  /** Loads values and counters into the histogram.
+    * This action can only be performed when there the histogram is empty
+    *
+    * @param counts Array of tuple of (value,count)
+    */
+  def loadData(counts: Array[(Double, Int)]): Unit = {
+    if (data.size > 0) {
+      throw new RuntimeException("Data can only be loaded during initialization")
+    }
+    counts.iterator.foreach(x => data.+=((x._1, x._2)))
+    require(checkSanity, "Invalid initialization from data")
+  }
+
+  /** Bin value access function
+    *
+    * @param bin bin number to access
+    * @return `v_bin` = value of bin
+    */
+  private[math] def getValue(bin: Int): Double = {
+    require(0 <= bin && bin < bins, bin + " not in [0, " + bins + ")")
+    data.apply(bin)._1
+  }
+
+  /** Returns the lower limit on values of the histogram
+    *
+    * @return lower limit on values
+    */
+  private def lower: Double = {
+    min
+  }
+
+  /** Returns the upper limit on values of the histogram
+    *
+    * @return upper limit on values
+    */
+  private def upper: Double = {
+    max
+  }
+
+  /** Checks whether the arraylist provided is properly sorted or not.
+    * All values should be in (lower,upper)
+    * All counters should be positive
+    *
+    */
+  private def checkSanity: Boolean = {
+    val size = bins
+    if (size == 0) return true
+    if (lower >= getValue(0)) return false
+    if (upper <= getValue(size - 1)) return false
+    for (i <- 0 to size - 2) {
+      if (getValue(i + 1) < getValue(i)) return false // equality will get merged later on
+      if (getCounter(i) <= 0) return false
+    }
+    if (getCounter(size - 1) <= 0) return false
+    // now bring the histogram to capacity and merge all <i>very-close</i> elements
+    while (mergeElements()) {}
```

Comment:
This seems a bit dangerous. What are the termination guarantees of mergeElements? Can we avoid the while statement here?